subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -750,3 +766,50 @@ int azure_fota_msg_process(const char *const buf, size_t size)
 
 	return 1;
 }
+
+void azure_fota_modem_firmware_update_validate(void)
+{
+	/* Handle return values relating to modem firmware update */
+	switch (modem_lib_init_result) {
+	case 0:
+		/* Initialization successful, no action required. */
+		return;
+	case MODEM_DFU_RESULT_OK:
+		LOG_DBG("MODEM UPDATE OK. Will run new modem firmware after reboot");
+		break;
+	case MODEM_DFU_RESULT_UUID_ERROR:
+	case MODEM_DFU_RESULT_AUTH_ERROR:
+		LOG_ERR("MODEM UPDATE ERROR %d. Will run old firmware", modem_lib_init_result);
+		break;
+	case MODEM_DFU_RESULT_HARDWARE_ERROR:
+	case MODEM_DFU_RESULT_INTERNAL_ERROR:
+		LOG_ERR("MODEM UPDATE FATAL ERROR %d. Modem failure", modem_lib_init_result);
+		break;
+	default:
+		/* All non-zero return codes other than DFU result codes are
+		 * considered irrecoverable and a reboot is needed.
+		 */
+		LOG_ERR("nRF modem lib initialization failed, error: %d", modem_lib_init_result);
+		break;
+	}
+
+	/* If the Carrier library is enabled, the reboot will be handled there. */
+	if (!IS_ENABLED(CONFIG_LWM2M_CARRIER)) {
+		LOG_DBG("Rebooting...");
+		LOG_PANIC();
+		sys_reboot(SYS_REBOOT_COLD);
+	}
```

Comment:
I wonder if this, or aspects of this routine, should be centralized under the FOTA library, especially since such a centralization is, AFAICT being removed from lwm2m

But that has pros and cons, so if you've already considered that and decided it is better to leave this fully on the shoulders of applications, then no worries

Reply:
@jayteemo I'd also be interested in your opinion on this, since maybe I am just being naive by even suggesting it, and you are a lot more familiar with the FOTA lib than I am

Reply:
what "FOTA library" are you referring to?

Reply:
Sorry, I was thinking in vague terms, there isn't actually FOTA library per say, but something like this could be a utility of `fota_download` or of the `dfu` lib

Reply:
Having it in FOTA download makes sense. But, as it stands only AWS and Azure FOTA will actually use this API.
LwM2M firmware reinitializes the modem after DFU (basically handles the init ret it during the FOTA process)
nRF Cloud have a dedicated API that needs to be called after reboot: `nrf_cloud_fota_pending_job_validate`.

We could look into generalizing this amongst the FOTA libraries down the road, but that would definitely require more effort.